### PR TITLE
Feature/codable keys

### PR DESCRIPTION
**Pull request by CaptainRedmuff**
This PR adds shortened coding key values to reduce the file size of meshes when saved to disc. 

Only encoding the polygon material if a value is present and shortening the coding key values yields a significant reduction in the file size for meshes.

Initial results for this PR have seen simple meshes compressed by about 30%

### CODE REVIEW

**Comment by CaptainRedmuff** [comment]
Is there any requirement for supporting encoding/decoding of structs such as `Vector` where components are expected to be supplied in varying formats? 

For example, `Vector` (and other structs) appear to offer support for decoding values where the `z` component is omitted presumably to facilitate using `Vector` as a 2d struct for texture coordinates.

Does this library intend to support exporting meshes for specific definitions such as COLLADA or .obj files or any other external file format? If this is not the case the implementations of `Codable` could be simplified further if keyed containers are used where the key/value pairs of the structs being parsed are in a known, fixed format.

Whilst omitting these properties does reduce the size of the overall output, the complexity added for parsing each struct is considerably more cumbersome than simply encoding these "empty" values as is.

I will fix the broken `CodingTests` to illustrate my thoughts but I would be interested on your input for simplifying the codables as above.

**Comment by nicklockwood** [comment]
Thanks for looking at this. There's no particular requirement to support importing/exporting standard formats. The aims for the Codable support were basically:

1) support saving and loading meshes generated using the Euclid API
2) support loading hand-written mesh data in whatever forms are most convenient for humans to write

I'm not a huge fan of using single-letter keys. They're less natural for humans to write, which is annoying for goal 2, and they also don't reduce the size of serialized output as much as other strategies such as unkeyed containers, which is the solution I've used for things like Vector. Making that change now would also break backwards compatibility with existing stored data.

I think the goal of reducing serialized mesh size might be better served by adding a new, keyless serialization format that is similar to traditional formats, with vertex, normal, texture and index data all stored in separate keyless arrays. This could then be the default format for saving, while still loading data in the current format if detected.

File: Sources/Angle.swift [hunk 1]
```diff
@@ -18,24 +18,18 @@ public struct Angle: Hashable, Comparable {
...
-        }
-        self = .degrees(try container.decode(Double.self, forKey: .degrees))
+        let container = try decoder.container(keyedBy: CodingKeys.self)
+        self = .radians(try container.decode(Double.self, forKey: .radians))
     }
 
     public func encode(to encoder: Encoder) throws {
-        try radians.encode(to: encoder)
+        var container = encoder.container(keyedBy: CodingKeys.self)
+        try container.encode(radians, forKey: .radians)
```
Comment (nicklockwood) [inline Sources/Angle.swift:32]: I don't understand this change. I thought the idea was to reduce the size of exported data, but this replaces an encoding as a simple number with an object containing a key, so:

0.5

becomes

{ "r": 0.5 }

File: Sources/Paths.swift [hunk 2]
```diff
@@ -38,21 +38,22 @@ public struct PathPoint: Hashable {
...
 
     public func encode(to encoder: Encoder) throws {
-        var container = encoder.unkeyedContainer()
-        try container.encode(position.x)
-        try container.encode(position.y)
-        try position.z == 0 ? () : container.encode(position.z)
-        try isCurved ? container.encode(true) : ()
+        var container = encoder.container(keyedBy: CodingKeys.self)
+        try container.encode(position, forKey: .position)
+        try container.encode(isCurved, forKey: .isCurved)
```
Comment (nicklockwood) [inline Sources/Paths.swift:56]: This again increases the size of the encoded format by replacing the unkeyed flat format:

[x, y, z, true]

with a nested object format:

{ "p": { "x": x, "y": y, "z": z }, "c": true }

File: Sources/Plane.swift [hunk 3]
```diff
@@ -45,30 +45,21 @@ public struct Plane: Hashable {
...
 
     public func encode(to encoder: Encoder) throws {
-        var container = encoder.unkeyedContainer()
-        try container.encode(normal.x)
-        try container.encode(normal.y)
-        try container.encode(normal.z)
-        try container.encode(w)
+        var container = encoder.container(keyedBy: CodingKeys.self)
+        try container.encode(normal, forKey: .normal)
+        try container.encode(w, forKey: .w)
```
Comment (nicklockwood) [inline Sources/Plane.swift:62]: This again increases the encoded size, replacing an unkeyed array of 4 numbers:

[x, y, z, w]

with an object:

{ "n": { "x": x, "y": y, "z": z }, "w": w }

File: Sources/Vertex.swift [hunk 4]
```diff
@@ -47,23 +47,25 @@ public struct Vertex: Hashable {
...
+        texcoord = try container.decode(Vector.self, forKey: .texcoord)
+        
     }
 
     public func encode(to encoder: Encoder) throws {
         var container = encoder.container(keyedBy: CodingKeys.self)
         try container.encode(position, forKey: .position)
         try container.encode(normal, forKey: .normal)
-        try texcoord == .zero ? () : container.encode(normal, forKey: .texcoord)
+        try container.encode(texcoord, forKey: .texcoord)
```
Comment (nicklockwood) [inline Sources/Vertex.swift:68]: Again, this seems to increase the size of untextured meshes for no obvious reason?

**Review by nicklockwood — COMMENTED** [review]
I don't really understand the goal here. The PR states that the intention is to reduce the serialized data size, but with a couple of exceptions all of these changes actually increase the size.

If the goal is to improve the speed of encoding/decoding, it would be helpful to get some metrics on how it affects that.

If the goal is just to simplify or normalize the way encoding is done, I think that's too hhigh of a price to pay for breaking backwards compatibility.

I don't see a way that I can accept this PR, but if you'd like to provide your sample mesh I believe I can make some changes that will dramatically reduce the encoded size without breaking compatibility with existing serialized data.

**Comment by CaptainRedmuff** [comment]
Agreed; this PR got a little out of hand and snowballed into something else entirely. 

The original aim was to reduce the file size of meshes but I realise now I was trying to achieve this the wrong way and for all the wrong reasons. 

I am building an app where models are saved/loaded in a hierarchical graph of meshes performing CSG operations at each node. Persisting the rotation and scale etc are needed here to be able to reproduce the model when loading from disc and Euclid does this perfectly as is.

What I now realise I require is an additional means to encode the final mesh as simplified array of polygon vertices, solely the position, normal and texture coordinates. My mistake here was assuming that the changes I had proposed would achieve this goal. 

Would you have any suggestions on options for exporting mesh data in a trimmed down, keyed format similar to [this](https://gist.github.com/CaptainRedmuff/d4b7eb037f50e637e2ff075ed1370b91)?

**Comment by nicklockwood** [comment]
You might also want to take a [ShapeScript](https://github.com/nicklockwood/ShapeScript), which includes a scene graph implementation built on top of Euclid. I'm thinking of moving some of that logic into Euclid's core, although that will mean decoupling it from SceneKit a bit.

**Pull request closed without merging**